### ISSUE TITLE
Homebrew Cask install fails

### ISSUE THREAD

**Issue by danieljimenez** [chef/chef#3022]
On some of my machines, homebrew::cask attempts to use macports to install cask. I can't find any consistency, except that this started happening on freshly installed Mac as of the other day. Relatively new to Chef, so any assistance would be appreciated.

```
 Recipe: homebrew::cask
   * homebrew_tap[caskroom/cask] action tap (up to date)
   * macports_package[brew-cask] action install
     
     ================================================================================
     Error executing action `install` on resource 'macports_package[brew-cask]'
     ================================================================================
     
     Errno::ENOENT
     -------------
     No such file or directory - port
     
     Resource Declaration:
     ---------------------
     # In /var/chef/cache/cookbooks/homebrew/recipes/cask.rb
     
      23: package 'brew-cask'
      24: 
     
     Compiled Resource:
     ------------------
     # Declared in /var/chef/cache/cookbooks/homebrew/recipes/cask.rb:23:in `from_file'
     
     macports_package("brew-cask") do
       action :install
       retries 0
       retry_delay 2
       default_guard_interpreter :default
       package_name "brew-cask"
       timeout 900
       declared_type :package
       cookbook_name "homebrew"
       recipe_name "cask"
     end
```

**Comment by danieljimenez** [comment]
This seems to be happening with other packages as well. Any idea why it wants to use port? I'm on Chef 12.

**Comment by jdmundrawala** [comment]
The exact chef client version and stack trace would be helpful. It should say in the logging where it dumped the stack to.

**Comment by danieljimenez** [comment]
Will share today or tomorrow.

**Comment by danieljimenez** [comment]
Stack trace:
https://gist.github.com/danieljimenez/437cad607857f9fb9504

I can also post the output from knife bootstrap if that would help.

**Comment by jgoldschrafe** [comment]
This is a regression in Chef 12.1.0 that does not occur in 12.0.3.

**Comment by danieljimenez** [comment]
Yikes! So this bug came in omnibus installer?

Daniel

**Comment by danieljimenez** [comment]
Thanks so much @jgoldschrafe.

**Comment by danieljimenez** [comment]
I concur with @jgoldschrafe, this is a regression. After installing 12.0.3 the same recipe worked great on freshly installed copy of Mac OS X.

**Comment by bcg62** [comment]
Also happening to me. Anyone working on a fix?

**Comment by lamont-granquist** [comment]
suspect this was broken in c793d2d5

**Comment by lamont-granquist** [comment]
its trying to use the provider_priority_map to resolve a duplicate resource, which isn't going to work.

**Comment by btm** [comment]
@bcg62 at the moment we think we're going to revert https://github.com/chef/chef/pull/2722 for a patch release and that feature will need to be rewritten some other way. Patch release expected Monday.

**Comment by jaymzh** [comment]
@lamont-granquist  why won't that work, theres:
```
 priority :package, Chef::Provider::Package::Homebrew, os: "darwin"
```

**Comment by lamont-granquist** [comment]
Well it doesn't actually work.  That priority is for the *provider*_priority_map and you're trying to resolve a conflict with two *resources* being selected.  So the provider_priority_map is never consulted.  What is happening is that based on class loading order one or the other of those are getting selected.  So I think what this patch did is effectively revert the change from macports to homebrew for everyone.

**Comment by jaymzh** [comment]
Ohhh I see. Wait, if we just revert the 'provides' in the resource, but leave the one in the provider, shouldn't that work?

**Comment by lamont-granquist** [comment]
i think not quite.  package will resolve to a homebrew_package (Chef::Resource::HomebrewPackage) which will get passed into the provider resolver and the homebrew package provider is the only provider which offers that one.

**Comment by lamont-granquist** [comment]
I'm spiking on making resource resolution look more like provider resolution right now.

**Comment by jaymzh** [comment]
hmm I see. The new resolution code scares me. :)

Thanks @lamont-granquist for jumping on this.

**Comment by btm** [comment]
Lamont's working on a feature for this over on https://github.com/chef/chef/pull/3077, which we're planning to include in 12.2.0 next week (if we can stay on schedule). I'll put together a revert of this for the 12.1.2 regression fix so we can unblock everyone else for now.

**Comment by jaymzh** [comment]
please don't break chef for us in 12.1.2 ...

**Comment by danieljimenez** [comment]
It's already broken in 12.1.2.

**Comment by jaymzh** [comment]
@danieljimenez only because they reverted our PR. WHich I'm expressing my disapproval of!

**Comment by btm** [comment]
Chef 12.0.0 changed the default provider for OS X from macports to homebrew. This decision was formalized by the community in [RFC016](https://github.com/chef/chef-rfc/blob/master/rfc016-homebrew-osx-package-provider.md).

Given a simple recipe of ```package "wget"```, this worked in 12.0.0:
```
# chef-apply _12.0.0_ /tmp/pack.rb 
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * package[wget] action install
    - install version 1.16.1 of package wget
```

In #2722 @jaymzh/@patcox wanted to add functionality to be able to upgrade to Chef 12  without having to modify every existing recipe from "package" to "macports_package".

However, in doing so, #2722 regressed RFC016 (implemented in #1709). The same recipe now tries to use macports again, which broke @danieljimenez, @jgoldschrafe, @bcg62, and certainly others.  

```
# chef-apply _12.1.0_ /tmp/pack.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * macports_package[wget] action install
 ```

#2722 is a regression caused by a new feature. The right thing to do is fix the regression for everyone (and ship), and then work on reimplementing the feature without the bug (and then ship again). In this case, @jaymzh, you should stay on 12.1.1 for a week until you and @lamont-granquist can mutually re-write your feature.

**Comment by lamont-granquist** [comment]
Your PR broke RFC016 and everyone else.  We really want to get that reverted as quickly as possible and keep homebrew the default package provider for Mac and not have more people pulling 12.1.x and writing incorrect code against RFC016 violating behavior and becoming bug dependent on that and breaking in 12.2.0.  In fact since we broke our documented API in 12.1.0 we have to revert that quickly according to SemVer.

**Comment by jtimberman** [comment]
I concur with @btm and @lamont-granquist: We should revert the change that introduced the bug and get a bugfix release out ASAP. We also need to ensure that we have test coverage to reduce the chance that we have this kind of regression in the future.

Let's plan on targeting 12.2.0 with that new feature with the bug fixed.

**Comment by jaymzh** [comment]
*sigh* Fine.

That said, I don't see it as my PR breaking a feature, I see 12.0.0 as providing no upgrade path and thus breaking *all* users who didn't start on Chef 12.

**Comment by lamont-granquist** [comment]
Nope you broke an RFC'd feature.

And I think you could still do this in a cookbook library file:

```
Chef::Resource::MacportsPackage.provides :package, os: "darwin"
Chef::Provider::Package::Macports.provides :package, os: "darwin"
```

That'll get evaluated after the resources and providers classes are initially evaluated and it'll overwrite the Chef::Resource.node_map and point the package provider at mac ports.

Protect that with some Gem::Version test on Chef::VERSION between 12.0.0 and 12.1.x and that should get you going.

**Comment by danieljimenez** [comment]
This aligns with semver. Major version number increase may break things. 

Daniel

**Comment by jaymzh** [comment]
I'm pretty sure we tried that and it didn't work. but whatever, we'll skip a version.

**Comment by btm** [comment]
Smoke testing on 12.1.2 looks good, should be releasing the package today.
```
# chef-apply /tmp/pack.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * homebrew_package[wget] action install
    - install version 1.16.1 of package wget
# chef-apply -v
Chef: 12.1.2
# 
```

**Comment by btm** [comment]
This should be fixed in [12.1.2](https://www.chef.io/blog/2015/03/23/chef-client-12-1-2-released/).

**Comment by legal90** [comment]
@lamont-granquist  @btm It is fixed in 12.1.2, but has been broken again in 12.2.1:

```
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/homebrew/recipes/cask.rb

            23: package 'brew-cask'
            24:

       Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/homebrew/recipes/cask.rb:23:in `from_file'

           macports_package("brew-cask") do
             action :install
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             package_name "brew-cask"
             timeout 900
             declared_type :package
             cookbook_name "homebrew"
             recipe_name "cask"
           end
```

**Comment by danieljimenez** [comment]
I just came across the same thing.

**Comment by bcg62** [comment]
Same case for me

**Comment by lozzd** [comment]
Seeing this as well for a simple brew install.

    [2015-04-23T21:18:13+01:00] INFO: Processing macports_package[pigz] action install
    
    ================================================================================
    Error executing action `install` on resource 'macports_package[pigz]'
    ================================================================================
    
    Errno::ENOENT
    -------------
    No such file or directory - port
    
    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/installerz/recipes/pigznhtop.rb
    
      1: package "pigz" do
      2:   action :install
      3: end
      4:
    
    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/installerz/recipes/pigznhtop.rb:1:in `from_file'
    
    macports_package("pigz") do
      action [:install]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      package_name "pigz"
      timeout 900
      declared_type :package
      cookbook_name "installerz"
      recipe_name "pigznhtop"
    end


    chef-client -v
    Chef: 12.2.1

**Comment by jonlives** [comment]
I'm reopening this issue as there have now been multiple reports of this being broken in 12.2.1 - @lamont-granquist @btm any idea if the fix was reverted in that version or something?

**Comment by jdmundrawala** [comment]
The fix was applied to the 12-stable branch, not master.
12.2 was not done of the 12-stable branch as we decided to stop doing long lived branches like that.
This problem should be resolved in 12.3

**Comment by btm** [comment]
@jonlives to add to @jdmundrawala's comment. The fix was in 12-stable, so fixed in 12.1.2, but then 12.2.x came out of a different branch that wasn't based on the 12-stable branch, and we accidentally lost the revert commit.

@jonlives Can you test the 12.3.0 RC? I think https://github.com/chef/chef/pull/3077 fixes this (where as the revert commit was a work around).

**Comment by btm** [comment]
I've confirmed this is fixed again in 12.3.0.

```
$ brew list pigz
Error: No such keg: /usr/local/Cellar/pigz

$ cat /tmp/pack.rb
package "wget"
package "pigz"

$ chef-apply -v
Chef: 12.3.0

$ sudo chef-apply /tmp/pack.rb
ffi-yajl/json_gem is deprecated, these monkeypatches will be dropped shortly
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * package[wget] action install (up to date)
  * package[pigz] action install
    - install version 2.3 of package pigz

$ brew list pigz
/usr/local/Cellar/pigz/2.3.3/bin/pigz
/usr/local/Cellar/pigz/2.3.3/bin/unpigz
/usr/local/Cellar/pigz/2.3.3/share/man/ (2 files)
```

I went back and also confirmed that 12.2.1 used macports (and it should not have) just to make sure I've got my head on straight this early in the morning.
